sklearn_numba_dpex/kmeans/kernels/kmeans_plusplus.py

```diff
@@ -0,0 +1,224 @@
+import math
+from functools import lru_cache
+
+import numpy as np
+import numba_dpex as dpex
+
+from ._base_kmeans_kernel_funcs import (
+    _make_initialize_window_kernel_funcs,
+    _make_accumulate_sum_of_ops_kernel_func,
+)
+from sklearn_numba_dpex.common.random import make_rand_uniform_kernel_func
+
+# NB: refer to the definition of the main lloyd function for a more comprehensive
+# inline commenting of the kernel.
+
+
+@lru_cache
+def make_init_kmeansplusplus_kernel(
+    n_samples,
+    n_features,
+    preferred_work_group_size_multiple,
+    work_group_size,
+    dtype,
+):
+
+    zero_idx = np.int64(0)
+    zero_init = dtype(0.0)
+
+    @dpex.kernel
+    # fmt: off
+    def init_kmeansplusplus(
+        X_t,                      # IN READ-ONLY   (n_features, n_samples)
+        sample_weight,            # IN READ-ONLY   (n_samples,)
+        centers_t,                # OUT            (n_features, n_clusters)
+        center_indices,           # OUT            (n_clusters,)
+        closest_dist_sq,          # OUT            (n_samples,)
+    ):
+    # fmt: on
+        sample_idx = dpex.get_global_id(zero_idx)
+        if sample_idx >= n_samples:
+            return
+
+        starting_center_id_ = center_indices[zero_idx]
+
+        sq_distance = zero_init
+        for feature_idx in range(n_features):
+            diff = X_t[feature_idx, sample_idx] - X_t[feature_idx, starting_center_id_]
+            sq_distance += diff * diff
+
+        sq_distance *= sample_weight[sample_idx]
```

Comment:
I don't have a source for that, I'd say we shouldn't so that a point that is included two times in the dataset has the same weight that a point with sample_weight = 2.

Reply:
WDYM about lloyd and elkan ? I think weights in lloyd / elkan should only be used in the update step, not when computing closest centroids, and it seems that our implementation and sklearn's agree on that ?

But I think the meaning of weights in kmeans++ is a bit different and there it makes sense to use it to weight distances,